Add reCaptcha support to Contact Form.

rahulsprajapati opened the pull request:
Fixes #840 

#### Changes proposed in this Pull Request:

* Move `recaptcha.php` file to the lib dir to support it for `Contact Form` module. Now we can use jetpack_require_lib for recaptcha.  
* ~Introduce new filter `grunion_contact_form_enable_recaptcha` to disable recaptcha for Contact form if a user only wants that for current share email ( as it is in the current version ).~
* Add recaptcha markup using `Jetpack_ReCaptcha` class in form and validate it before processing.

#### Testing instructions:

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.

Would you like this feature to be tested by Beta testers as well?
Please add instructions to to-test.md in a new commit as part of your PR.
-->

* Follow this instruction to setup reCaptcha https://jetpack.com/tag/recaptcha/
* It will enable the reCaptcha on Contact Form by default after above configuration.
* You can add Contact Form by following instruction from here: https://jetpack.com/support/contact-form/

<!-- Add the following only if this is meant to be in changelog -->
#### Proposed changelog entry for your changes:
* Contact Form: Add reCaptcha Support

rahulsprajapati committed:
Add reCaptcha support to Contact Form.

jeherve reviewed (commented):
Instead of moving reCaptcha to the modules level, I would recommend moving it with the other libraries here?
https://github.com/Automattic/jetpack/tree/master/_inc/lib

You can then rely on `jetpack_require_lib` to load the library anywhere.

rahulsprajapati committed:
Move 'recaptcha.php' to lib dir and use `jetpack_require_lib` to load. Use Jetpack_Constants::get_constant to access constant value.

rahulsprajapati committed:
Update phpunit for recaptcha test case.

rahulsprajapati commented:
Hey, @jeherve 
I've updated the PR and moved the reCaptcha to lib. I've two concern about this PR, was hoping you can help here.

1. I've added `grunion_contact_form_enable_recaptcha` filter with `@since 6.1-alpha` I think that should be updated with latest tag/version when/if it's merged for release. 
2. I've kept this `reCaptcha` enabled by default and provided to disable it via the above filter, do you think we should do opposite? Since currently, some sites will have reCaptcha enabled for email-share and when this update release reCaptcha will be enabled by-default and user might not have any idea or do not want it to Contact Form ( I'm not sure if the user even want that but just had a thought ) and if so user have to disable it via this filter. So maybe we can add an option to enable it via the filter and disable it by default. 

Let me know your thought on this.

rahulsprajapati commented:
I think some thing wrong with test cases. It failed for two travis-ci job:
https://travis-ci.org/Automattic/jetpack/jobs/370256583
https://travis-ci.org/Automattic/jetpack/jobs/370256585
because of same error:

> 1) WP_Test_Jetpack_Sync_Sender::test_delays_next_send_if_exceeded_sync_wait_threshold
Failed asserting that false is true.

Which I think is not related to this PR specific. And its happening for other recent PR/travis-ci job.
ref: https://travis-ci.org/Automattic/jetpack/jobs/370180488
https://travis-ci.org/Automattic/jetpack/jobs/370196894

dereksmart commented:
> Which I think is not related to this PR specific

@rahulsprajapati probably correct there. That test is pretty fickle. I restarted the tests.

rahulsprajapati commented:
Thanks, @dereksmart for restart the tests, all tests are passed this time.

zinigor reviewed (dismissed):
Works great, code looks good as well, thank you!

rahulsprajapati commented:
>This PR needs a rebase now

Done. Resolved Conflict.

dereksmart commented:
I'm thinking that this should not be automatically enabled if Recaptcha keys are set, and instead would be a better form field to add per form, or a global setting that someone must enable.  

The biggest concern I have is syncing this back to wpcom, where there is a key for all of wpcom sites, and enabling it there would enable it globally without users knowing about it. It changes the front-end flow of their sites unexpectedly.  

Thoughts, anyone?

oskosk commented on modules/contact-form/grunion-contact-form.php:
```diff
@@ -288,6 +289,49 @@ function process_form_submission() {
+		 *
+		 * @module contact-form
+		 *
+		 * @since 6.1-alpha
```
This would need to be update to at least 6.2

oskosk commented on modules/contact-form/grunion-contact-form.php:
```diff
@@ -1756,8 +1800,49 @@ static function parse( $attributes, $content ) {
+			 *
+			 * @module contact-form
+			 *
+			 * @since 6.1
```
Same here. Please update to 6.2 at least

keoshi commented:
@rahulsprajapati Hi there, mind taking another look at the previous feedback from @dereksmart and @oskosk, so we can move this PR along please? Thank you in advance!

rahulsprajapati commented:
yeah sure @keoshi I'll update this PR ASAP.

@dereksmart 
I agree with you we should keep it disable and add some option to enable it. Can you suggest how/where we should give this option? Maybe we can keep it disable by-default by keeping `GRUNION_RECAPTCHA_ENABLED` constant false and user can enable by setting `GRUNION_RECAPTCHA_ENABLED` constant true or maybe can provide some option in Jetpack Settings page. 
ref: https://github.com/Automattic/jetpack/pull/9389#issuecomment-383701637

P.s: If `RECAPTCHA_PUBLIC_KEY` and `RECAPTCHA_PRIVATE_KEY` constant are not available/empty then it will be disabled since w/o it won't work anyway.

keoshi commented:
Thanks, @rahulsprajapati !

rahulsprajapati committed:
Update @since version for grunion_contact_form_enable_recaptcha filter.

jeherve commented:
> Maybe we can keep it disable by-default by keeping GRUNION_RECAPTCHA_ENABLED constant false and user can enable by setting GRUNION_RECAPTCHA_ENABLED constant true

This (or the `grunion_contact_form_enable_recaptcha` filter) seems like a good idea; I would go with that. 👍

rahulsprajapati committed:
Disable reCaptcha for contact form by-default. And remove grunion_contact_form_enable_recaptcha filter.

rahulsprajapati commented:
@jeherve 
I've updated this PR and kept recatcha disable by-default. It can be enabled by `GRUNION_RECAPTCHA_ENABLED` define this constant as true. 
And since this recaptcha will require `RECAPTCHA_PUBLIC_KEY` and `RECAPTCHA_PRIVATE_KEY` keys I've put that also in enable/disable flag. So, all three constants will be required to activate this feature. https://github.com/Automattic/jetpack/pull/9389/commits/8caa8f4deeab17226c4bca60441afebea1fb259b#diff-ba2abc1889d79bcf7293e6748302450bR298

```
define( 'GRUNION_RECAPTCHA_ENABLED', true );
define( 'RECAPTCHA_PUBLIC_KEY', '******' );
define( 'RECAPTCHA_PRIVATE_KEY', '*****' );
```

P.s.: removed `grunion_contact_form_enable_recaptcha` filter since that is no longer required. 

cc: @dereksmart

brbrr reviewed (dismissed):
I give it a shot, and it works great!

One thing I noticed, which actually confused me - reCaptcha works without defined `RECAPTCHA_PRIVATE_KEY` e.g. I can submit a form with success response.

Also, as mentioned above, all the `@since` docs should be updated to the next release which is 6.7 (hopefully we'll land this PR soon)

kraftbj commented:
@oskosk Since you requested changes previously, can you give this a look to 👍 or 👎 ?

kraftbj committed:
Update docblock @since

kraftbj reviewed (dismissed):
I didn't test it, but approving based on @brbrr's review. I updated the docblock, which led GitHub to dismiss it.

jmmeijer commented:
@rahulsprajapati Is this PR hanging on a version number in the comments?

rahulsprajapati commented:
hey @jmmeijer 

I don't think so this PR is hanging for the version number changes. Last @kraftbj have made changes to update `Since` for version 6.7.0, I'm not sure why this PR is on hold.

I can update it to 7.0.1 if we are plannign to release it in 7.0.1.

jeherve commented:
Sorry about that! This PR won't make it to Jetpack 7.0.1, but we'll see that it gets a last review and gets merged if possible for the Jetpack 7.1 release!

rahulsprajapati commented:
cool. No prob. @jeherve 🙂

jeherve committed:
Sharing: add placeholder file, to avoid fatals on update

jeherve committed:
Update docblock since

jeherve committed:
Update reCAPTCHA strings.

jeherve committed:
Use the Jetpack_Constants class to retrieve and check constants

jeherve commented:
Internal reference: D24787-code

jeherve reviewed (approved):
This looks good and works well in my tests. Merging.

jeherve commented:
Coming back to this. After further discussion, we will not be adding the feature to Jetpack right now. It will be removed in #11432.

jessefriedman commented:
@rahulsprajapati 

Thank you for investing the time to build this added functionality to the Jetpack contact form.  I really appreciate your commitment to reducing form spam and providing a positive WordPress site building experience.  

That being said I have decided to pause on including this code in Jetpack 7.1 in favor of an Akismet solution.  I am sorry I was not able to get involved earlier, that is my fault.  

As I'm sure you know Akismet is an Automattic made product that works really well with the Jetpack form to reduce spam.  As our focus is on supporting and integrating our own products first, it made sense to reduce the user choices here and promote Akismet as the primary solution.  

I am happy to talk more about this here or in email if you'd prefer.  My personal email address is jesse _at_ jetpack.com and I'm more than happy to answer any questions or hear any feedback from you. 

I really hope you decide to launch this code as its own plugin, one we could recommend if a user is specifically looking for a captcha solution to spam.  Please let me know if I can be of any help in the future with that, or anything else.  I really hope this doesn't discourage you from contributing in the future. 

Jesse

rahulsprajapati commented:
No worries @jessefriedman 🙂
Thank you for writing me back and explaination on why we are reverting this feature. It make sense to use/promote Automattic's own product Akismet which serve the purpose of reducing form spam/bot. 👍 

---

I just had a requirement to keep reCaptcha in Jetpack Contact Form. So was looking for some hooks to build a plugin but it seems we don't have a way to inject form body.

https://github.com/Automattic/jetpack/blob/305f967b3bc4f4d9c9e7328565f6aca5559d43bc/modules/contact-form/grunion-contact-form.php#L2035-L2037

I think maybe it's because we don't want unnessaory data in in contact form. But Let me know if we can add after/before action hook in contact form so I can write a plugin. 

---

For my requirement I used one workaround earlier and build a custom shortcode like: `custom-recaptcha` and editors will have to inject that in to content which worked for me. 

```php
[contact-form submit_button_text='Submit &gt;' to="test@test.com" subject="Connect With Us"]
[contact-field label="Name" type="name" required="1"]
[contact-field label="Email" type="email" required="1"]
[contact-field label="Comments" type="textarea"]
[custom-recaptcha]
[/contact-form]
```

jessefriedman commented:
Thank you for your understanding Rahul.  I will get with the team and get back to you on those hooks.

jeherve commented:
@rahulsprajapati I'd recommend checking this plugin to get some ideas on how to implement reCAPTCHA from a third-party plugin:
https://wordpress.org/plugins/recaptcha-for-jetpack/
https://github.com/DevriX/jetpack-recaptcha

You'll see that they are leveraging an existing filter, `jetpack_contact_form_is_spam`, that could be helpful for you as well.

That said, if you find that another filter may be useful somewhere else, don't hesitate to open a PR with your suggestion, we'll be happy to review it!

Bachsau commented:
You favor a privacy nightmare above a working solution. So it's just "not invented here".

alexandernst commented:
[This](https://github.com/Automattic/jetpack/pull/9389#issuecomment-467976310) makes a lot of sense from a business point of view (you're attracting/forcing users into your own, paid, solution). But the fact that Akismet is a paid product, and you're rejecting a free (for the users) alternative, coming from your own community, in the form a free (as in beer and as in speech) PR, makes this a completely dick move.

tt33tt commented:
Hi all, I want to implement reCAPTCHA in the Jetpack-Form. 
We can not using Akisemet for privacy issues in Germany. 
I used https://jetpack.com/support/sharing/#captcha but there is no CAPTCHA appearing. Do I need to do more?
The plugin https://wordpress.org/plugins/recaptcha-for-jetpack/ is not allowed anymore and not usable anymore.
Maybe I can not use the Jetpack-Plugin and need to use another forms plugin?

jeherve commented:
> The plugin wordpress.org/plugins/recaptcha-for-jetpack is not allowed anymore and not usable anymore.

There seems to be a copy of the plugin here:
https://github.com/DevriX/jetpack-recaptcha

I would recommend giving that a try.

> I used jetpack.com/support/sharing/#captcha but there is no CAPTCHA appearing.

This setting is useful when using the email sharing button. It does not apply to the contact form.
